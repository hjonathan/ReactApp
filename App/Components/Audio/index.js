import React, {Component} from "react";
import {Platform, View, Text, TouchableOpacity, ProgressBarAndroid} from "react-native";
import moment from "moment";
import Sound from "react-native-sound";
import AudioRecorder from "react-native-audio-record";
import MaterialIcon from "react-native-vector-icons/MaterialIcons";
import I18n from "../../I18n";
import * as Res from "../../Assets/resources";
import Styles from "./styles";

const PREFIX = "formsliderAud";
let timer = null;
/**
 * Class Audio
 * Example:
 *      ...
 *      import Audio from "../../Components/Audio";
 *
 *      class SomeComponent extends component {
 *          render () {
 *              return (
 *                  <Audio
 *                      onClose={()=>{}} // close function
 *                      saveAudio={() => {}} // saving function
 *                  />
 *              );
 *          }
 *      }
 */
class Audio extends Component {
    /**
     * Constructor
     * @param props
     */
    constructor (props) {
        super(props);
        this.state = {
            timer: null,
            minutes: "00",
            seconds: "00",
            time: 0.00,
            audioPath: null,
            stopped: false,
            loading: false,
            recording: false
        };
    }

    /**
     * Initializes audio options
     */
    componentDidMount () {
        const options = {
            sampleRate: 16000,
            channels: 1,
            bitsPerSample: 16,
            wavFile: this.nameGenerator()
        };
        AudioRecorder.init(options);
    }
    /**
     * Closes control panel
     */
    onClosePanel () {
        this.setState({
            time: 0,
            audioPath: null,
            loading: false
        });
        this.props.onClose({status: false});
    }

    /**
     * Generates a file name
     * @returns {string}
     */
    nameGenerator () {
        let date = moment().format("YYYYMMDD"),
            timestamp = moment().valueOf(),
            extension = "m4a",
            uid;
        uid = `${PREFIX}_${date}_${timestamp}.${extension}`;
        return uid;
    }

    /**
     * Records audio
     * @private
     */
    _record () {
        if (!this.state.recording && !this.state.loading) {
            this.timeStart();
            AudioRecorder.start();
            this.setState({
                stopped: false,
                recording: true,
                loading: true
            });
        }
    }

    /**
     * Stops the current recording
     * @private
     */
    async _stop () {
        if (this.state.recording) {
            this.timeClear();
            let audioFile = await AudioRecorder.stop();
            this.setState({
                stopped: true,
                recording: false,
                loading: false,
                audioPath: audioFile
            });
        }
    }

    /**
     * Plays the current record
     * @private
     */
    _play () {
        if (this.state.stopped && !this.state.recording) {
            const sound = new Sound(this.state.audioPath, null, (error) => {
                if (error) {
                    console.error(error);
                    return;
                }
                this.setState({loading: true});
                this.timeStart();
                sound.play((success) => {
                    this.setState({loading: !success});
                    this.timeClear();
                });
            });
        }
    }

    /**
     * Saves current record
     * @private
     */
    _save () {
        if (this.state.stopped && this.state.audioPath) {
            this.props.saveAudio({path: this.state.audioPath});
        }
    }

    /**
     * Start time of chronometer
     */
    timeStart = () => {
        timer = setInterval(() => {
            let {minutes} = this.state,
                seconds = (Number(this.state.seconds) + 1).toString();
            if (Number(this.state.seconds) === 59) {
                minutes = (Number(this.state.minutes) + 1).toString();
                seconds = "00";
            }
            this.setState({
                minutes: minutes.length === 1 ? `0${minutes}` : minutes,
                seconds: seconds.length === 1 ? `0${seconds}` : seconds
            });
        }, 1000);
        this.setState({timer});
    };

    /**
     * End time of chronometer
     */
    timeClear () {
        this.setState({
            timer: null,
            minutes: "00",
            seconds: "00"
        });
        clearInterval(timer);
    }

    /**
     * Renders progress bar
     * @returns {*}
     */
    progressBar () {
        if (Platform.OS === "android") {
            return (
                <ProgressBarAndroid styleAttr="Horizontal" color={Res.colors.teal} />
            );
        }
        return null;
    }

    /**
     * Renders header
     * @returns {*}
     */
    renderHeader () {
        return (
            <View style={Styles.headerContainer}>
                <View style={Styles.headerTitle}>
                    <Text style={Styles.textTitle}>
                        {I18n.t("selectAudioRecordDialogtitle")}
                    </Text>
                </View>
                <TouchableOpacity
                    onPress={() => {
                        this.onClosePanel();
                    }}
                >
                    <MaterialIcon name="clear" style={Styles.closeIcon} />
                </TouchableOpacity>
            </View>
        );
    }

    /**
     * Renders indeterminate progressBar and timer
     * @returns {*}
     */
    renderProgressBar () {
        if (this.state.loading) {
            return (
                <View style={{backgroundColor: Res.colors.white}}>
                    {this.progressBar()}
                    <View>
                        <Text>{`${this.state.minutes}:${this.state.seconds}`}</Text>
                    </View>
                </View>
            );
        }
        return null;
    }

    /**
     * Renders record button
     * @returns {*}
     */
    renderRecordButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this._record();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="fiber-manual-record" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("recordAudioStringDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders stop button
     * @returns {*}
     */
    renderStopButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this._stop();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="stop" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("stopStringAudioDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders play button
     * @returns {*}
     */
    renderPlayButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this._play();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="play-arrow" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("playStringAudioDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders save button
     * @returns {*}
     */
    renderSaveButton () {
        return (
            <View style={[Styles.buttonContainer, {margin: 4}]}>
                <TouchableOpacity
                    onPress={() => {
                        this._save();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="send" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("savedStringAudioDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders component
     * @returns {*}
     */
    render () {
        return (
            <View style={Styles.mainContainer}>
                <View style={Styles.panelContainer}>
                    {this.renderHeader()}
                    {this.renderProgressBar()}
                    <View style={Styles.subContainer}>
                        {this.renderRecordButton()}
                        {this.renderStopButton()}
                        {this.renderPlayButton()}
                        {this.renderSaveButton()}
                    </View>
                </View>
            </View>
        );
    }
}

export default Audio;
