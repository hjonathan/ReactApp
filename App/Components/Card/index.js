import React, {Component} from 'react';
import {View} from 'react-native';
import Styles from './styles';

/**
 * Class Card
 * Example:
 * import Card from '../Card';
 * render() {
 *      return(
 *          <Card>
 *              {content}
 *          </Card>
 *      ),
 * }
 */
class CardView extends Component {
    /**
     * Constructor cardView
     * @param props
     */
    constructor(props) {
        super(props);
    };

    /**
     * Render
     * @param data
     * @returns {*}
     */
    render() {
        return (
            <View style={Styles.mainContainer} elevation={3}>
                {this.props.children}
            </View>
        );
    }
}

export default CardView;