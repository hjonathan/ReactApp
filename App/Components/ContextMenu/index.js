import React, {Component} from "react";
import {View, Text, TouchableOpacity} from "react-native";
import MaterialIcon from "react-native-vector-icons/MaterialIcons";
import I18n from "../../I18n";
import Styles from "./styles";

/**
 * Context menu class
 * Example:
 *      ...
 *      import ContextMenu from "../../Components/ContextMenu";
 *
 *      class SomeComponent extends component {
 *          render () {
 *              return (
 *                  <ContextMenu
 *                      type="audio" // audio \\ image \\ video
 *                      onClose={()=>{}} // close function
 *                      takeOne={() => {}} // creates a file function
 *                      selectFile={() => {}} // access to gallery
 *                  />
 *              );
 *          }
 *      }
 */
class ContextMenu extends Component {
    /**
     * Closes view
     */
    onCloseMenu () {
        if (this.props.onClose) {
            this.props.onClose({status: false});
        }
    }

    /**
     * Executes takeOne function
     */
    takeOne () {
        if (this.props.takeOne) {
            this.props.takeOne({status: true});
        }
    }

    /**
     * Executes selectFile function
     */
    selectFile () {
        if (this.props.selectFile) {
            this.props.selectFile();
        }
    }

    /**
     * Renders header
     * @returns {*}
     */
    renderHeader () {
        return (
            <View style={Styles.headerContainer}>
                <View style={Styles.headerTitle}>
                    <Text style={Styles.headerText}>
                        {I18n.t("selectAudioDialogtitle")}
                    </Text>
                </View>
                <TouchableOpacity
                    onPress={() => {
                        this.onCloseMenu();
                    }}
                >
                    <MaterialIcon name="clear" style={Styles.closeIcon} />
                </TouchableOpacity>
            </View>
        );
    }

    /**
     * Renders take one button
     * @returns {*}
     */
    renderTakeButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this.takeOne();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="mic" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <View style={Styles.labelButton}>
                    <Text style={{textAlign: "center"}}>{I18n.t("selectAudioStringDialog")}</Text>
                </View>
            </View>
        );
    }

    /**
     * Renders gallery button
     * @returns {*}
     */
    renderGalleryButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this.selectFile();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="graphic-eq" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <View style={Styles.labelButton}>
                    <Text style={{textAlign: "center"}}>{I18n.t("selectGalleryStringAudioDialog")}</Text>
                </View>
            </View>
        );
    }

    /**
     * Renders component
     * @returns {*}
     */
    render () {
        return (
            <View style={Styles.mainContainer}>
                <View style={Styles.subContainer}>
                    {this.renderHeader()}
                    <View style={Styles.panelContainer}>
                        {this.renderTakeButton()}
                        {this.renderGalleryButton()}
                    </View>
                </View>
            </View>
        );
    }
}

export default ContextMenu;
