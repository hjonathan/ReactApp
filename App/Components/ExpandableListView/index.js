import React, {Component} from "react";
import {Text, View, Image, TouchableHighlight, Animated} from "react-native";
import * as Res from "../../Assets/resources";
import Styles from "./styles";

class ListView extends Component {
    constructor (props) {
        super(props);

        this.icons = {
            up: Res.images.arrowHeadUp,
            down: Res.images.arrowHeadDown
        };

        this.state = {
            title: props.title,
            expanded: true
        };
        this.toggle = this.toggle.bind(this);
    }

    /**
     * Change status expanded for Content
     */
    toggle () {
        this.setState({
            expanded: !this.state.expanded
        });
    }

    /**
     * Renders
     * @returns {*}
     */
    render () {
        let icon = this.icons.down;

        if (this.state.expanded) {
            icon = this.icons.up;
        }

        return (
            <Animated.View style={Styles.container}>
                <View style={Styles.titleContainer}>
                    <Text style={Styles.title}>{this.state.title}</Text>
                    <TouchableHighlight
                        style={Styles.button}
                        onPress={this.toggle}
                        underlayColor="#f1f1f1"
                    >
                        <Image
                            style={Styles.buttonImage}
                            source={icon}
                        />
                    </TouchableHighlight>
                </View>
                {
                    this.state.expanded && (
                        <View style={Styles.body}>
                            {this.props.children}
                        </View>
                    )
                }
            </Animated.View>
        );
    }
}

export default ListView;
