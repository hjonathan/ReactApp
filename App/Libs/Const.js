import RNFS from "react-native-fs";

export const CASE = {
    INBOX: "Inbox",
    DRAFT: "Draft",
    PARTICIPATED: "Participated",
    UNASSIGNED: "Unassigned",
    STATUS: {
        NONE: "none",
        SENDING: "sending",
        WORKING: "working",
        OFFLINE: "offline"
    }
};

export const DATETIME = {
    LOCAL: "YYYY-MM-DD HH:mm:ss",
    LOCAL_UTC: "YYYY-MM-DD HH:mm:ss Z"
};

export const DATE = {
    SIMPLE: "YYYYMMDD",
    LOCAL: "YYYY-MM-DD"
};

export const TIME = {
    LOCAL: "HH:mm:ss"
};

export const DS = "/";

export const PATH = {
    PROJECT: `${RNFS.DocumentDirectoryPath}`,
    FORMRENDER: `${DS}FormRender`,
    STEPS: `${DS}Steps`,
    NEXTSTEPS: `${DS}NextSteps`,
    FROMS: `${DS}Forms`,
    DATA: `${DS}Data`,
    EXTERNALLIBS: `${DS}ExternalLibs`,
    BUILDPROD: `${DS}build-prod`,
    MUSIC: "/storage/emulated/0/Music/processmakeraudiorecord/"
};

export const FILENAMES = {
    EXTERNALMAP: "ExtMap.json"
};
