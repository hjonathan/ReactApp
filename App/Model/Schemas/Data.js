import moment from "moment";
import {CASE, DATETIME} from "../../Libs/Const";
/**
 * Schema for model Data
 */
export default class DataSchema {
    static schema = {
        name: "DataSchema",
        primaryKey: "caseId",
        properties: {
            caseId: {type: "string"},
            caseIndex: {type: "int"},
            data: {type: "string"},
            formId: {type: "string?", default: ""},
            stepId: {type: "string?", default: ""},
            stepHistory: {type: "string?", default: "[]"},
            createDate: {type: "string?", default: moment().format(DATETIME.LOCAL)},
            updateDate: {type: "string?", default: moment().format(DATETIME.LOCAL)},
            type: {type: "string?", default: ""},
            status: {type: "string?", default: CASE.STATUS.NONE}
        }
    };

    /**
     * Returns the data in json format
     * @returns {{}}
     */
    get dataJSON () {
        return this.data ? JSON.parse(this.data) : {};
    }

    /**
     * Returns the data in json format
     * @returns {Array}
     */
    get stepHistoryArray () {
        return this.stepHistory ? JSON.parse(this.stepHistory) : [];
    }
}
