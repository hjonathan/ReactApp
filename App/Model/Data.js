import _ from "lodash";
import moment from "moment";
import BaseModel from "./Interface/BaseModel";
import {RealmObject, DataSchema} from "./Schemas";
import {CASE, DATETIME} from "../Libs/Const";

/**
 * Model for Data
 */
class Data extends BaseModel {
    constructor () {
        super(DataSchema.schema.name);
    }

    /**
     * Save the list when the action formData/server/update is executed
     * @param params
     */
    saveData (params) {
        let saved = null,
            date = moment().format(DATETIME.LOCAL),
            data;
        RealmObject.write(() => {
            let oldData = {},
                dbCaseData = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${params.app_uid}"`);
            if (dbCaseData.length > 0) {
                oldData = _.values(dbCaseData)[0].dataJSON;
            }
            data = {
                caseId: params.app_uid,
                caseIndex: Number(params.del_index),
                formId: params.dyn_uid,
                data: JSON.stringify(Object.assign({}, oldData, params.data)),
                createDate: moment(params.createDate).format(DATETIME.LOCAL) || date,
                updateDate: date,
                status: CASE.STATUS.OFFLINE
            };
            saved = RealmObject.create(DataSchema.schema.name, data, true);
        });
        return {saved};
    }

    /**
     * Save case data when the action cases/data/request is executed
     * @param payload
     * @param state
     * @returns {{saved: *}}
     */
    saveCaseData (payload, state) {
        const {params, response} = payload;
        let saved = null,
            date = moment().format(DATETIME.LOCAL),
            data;
        RealmObject.write(() => {
            let oldData = {},
                dbCaseData = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${params.app_uid}"`);
            if (dbCaseData.length > 0) {
                oldData = _.values(dbCaseData)[0].dataJSON;
            }
            data = {
                caseId: params.app_uid,
                caseIndex: Number(params.del_index),
                data: JSON.stringify(Object.assign({}, oldData, response)),
                formId: params.formId || "",
                stepId: params.stepId || "",
                stepHistory: JSON.stringify(params.stepHistory) || "[]",
                createDate: moment(params.createDate).format(DATETIME.LOCAL) || date,
                updateDate: date,
                type: params.type || state.Cases.routeName,
                status: params.status || CASE.STATUS.NONE
            };
            saved = RealmObject.create(DataSchema.schema.name, data, true);
        });
        return {saved};
    }

    /**
     * Get data
     * @param caseId
     * @returns {*}
     */
    getDataCase (caseId) {
        return _.values(RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${caseId}"`)).shift();
    }

    /**
     * Get all data in status offline
     * @param caseId
     * @returns {*}
     */
    getAllCaseOffline () {
        return _.values(RealmObject.objects(DataSchema.schema.name).filtered(`status = "${CASE.STATUS.OFFLINE}"`));
    }

    /**
     * Get data of case with caseId and status is equal to working (work in progress)
     * @param caseId
     * @returns {*}
     */
    filterCaseIdWIP (caseId) {
        const result = RealmObject.objects(DataSchema.schema.name)
            .filtered(`caseId = "${caseId}" AND status = "${CASE.STATUS.WORKING}"`);
        return _.values(result.map((x) => {
            let newObject = _.assign({}, x);
            newObject.stepHistory = x.stepHistoryArray;
            newObject.data = x.dataJSON;
            return newObject;
        })).shift();
    }

    /**
     * Change status of row
     * @param caseId
     * @param status
     */
    changeStatus (caseId, status) {
        let dbData;
        RealmObject.write(() => {
            dbData = _.values(RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${caseId}"`)).shift();
            if (dbData) {
                dbData.status = status;
                RealmObject.create(DataSchema.schema.name, dbData, true);
            }
        });
    }

    /**
     * Remove the specified data case of storage.
     * @param value
     */
    destroyDataCase (value) {
        RealmObject.write(() => {
            let data = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${value}"`);
            RealmObject.delete(data);
        });
    }
    /**
     * Delete multiple Form
     */
    destroyObject () {
        RealmObject.write(() => {
            let allObjects = RealmObject.objects(DataSchema.schema.name);
            RealmObject.delete(allObjects); // Deletes all Forms
        });
    }
}

export default new Data();
