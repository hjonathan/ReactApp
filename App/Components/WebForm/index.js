import React, {Component} from "react";
import {WebView} from "react-native";
import RNFS, {DocumentDirectoryPath} from "react-native-fs";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import compatJs from "../../CompatPM/FormRender/JsInterface";
import Utils from "../../Utils";
import Actions from "../../Actions";
import Data from "../../Model/Data";
import {FILENAMES} from "../../Libs/Const";

const folderForm = "FormRender",
    folderBuild = "build-prod";

/**
 * Class FormRender
 */
export class WebForm extends Component {
    constructor (props) {
        super(props);
        this.state = Object.assign({}, props.payload, {
            webView: null,
            isLoad: false
        });
    }

    /**
     * Component did mount
     */
    componentDidMount () {
        this.props.spreadWebView(this.webView);
    }

    /**
     * Component will receive nextProps
     */
    componentWillReceiveProps (nextProps) {
        if (nextProps.errorNextStep) {
            this.webView.injectJavaScript(Utils.injectHideMaskLoading());
            this.props.clearWebForm();
        }
    }

    /**
     * Component update Handler
     * @param {*} nextProps
     * @param {*} nextState
     */
    componentDidUpdate (nextProps, nextState) {
        switch (this.props.flowStatus) {
            case "continueStep":
                this.webView.injectJavaScript(Utils.injectContinueStep({
                    path: this.props.pathNextStep
                }));
                break;
            case "loadForm":
                this.webView.injectJavaScript(Utils.injectLoadForm({
                    path: this.props.formPath
                }));
                break;
            case "hideFieldLoading":
                this.webView.injectJavaScript(Utils.injectHideFieldLoading(this.props.fileInput));
                break;
            case "setFormData":
                this.webView.injectJavaScript(Utils.injectSetFormData({
                    path: this.props.formDataPath,
                    stepData: this.props.stepData || {}
                }));
                break;
            case "setFiles":
                this.webView.injectJavaScript(Utils.injectSetFiles(this.props.fileUploaded));
                break;
            case "setFileVersions":
                this.webView.injectJavaScript(Utils.injectSetFileVersions(this.props.fileVersions));
                break;
            case "setLocation":
                this.webView.injectJavaScript(Utils.injectSetLocation(this.props.location));
                break;
            case "getFormData":
                this.webView.injectJavaScript(Utils.injectGetFormData());
                break;
            case "setCacheLibraryMap":
                this.webView.injectJavaScript(Utils.injectSetCacheLibraryMap({
                    path: `${FILENAMES.EXTERNALMAP}`
                }));
                break;
            default:
                break;
        }
    }

    /**
     * Listener on load Web View
     */
    onLoad = () => {
        if (!this.state.isLoad) {
            const {state} = this.props,
                {payload} = this.props;
            let str,
                path,
                caseData,
                data = {
                    steps: this.state.stepsPath,
                    isRTL: false,
                    offLineMode: false,
                    statusConnection: state.Net.isConnected
                },
                params = Object.assign({}, data, payload, Utils.mergePropsStateApi(state));
            caseData = Data.filterCaseIdWIP(params.app_uid);
            if (caseData) {
                path = `${RNFS.DocumentDirectoryPath}/data.json`;
                // TODO writing of the files will be resolved in the ticket PRE-101
                if (RNFS.exists(path)) {
                    RNFS.unlink(path);
                }
                RNFS.writeFile(path, JSON.stringify(caseData), "utf8")
                    .then(success => success)
                    .catch(error => error);
                params.dataPath = path;
                str = Utils.injectOpenStep(params);
            } else {
                str = Utils.injectOpenCase(params);
            }
            this.webView.injectJavaScript(`(${compatJs.toString()})()`);
            this.webView.injectJavaScript(str);
            this.props.finishLoadWebView();
            this.setState({isLoad: true});
        }
    };

    /**
     * Listener to message from webView
     * @param event
     */
    onMessage = (event) => {
        if (event && event.nativeEvent && event.nativeEvent.data) {
            const {type, data} = JSON.parse(event.nativeEvent.data),
                config = this.props.payload;
            this.props.onMessage({type, stepData: data, config});
        }
    };

    /**
     * Render the component
     * @returns {*}
     */
    render () {
        return (
            <WebView
                userAgent="formslider-android"
                ref={x => (this.webView = x)}
                source={{uri: `file://${DocumentDirectoryPath}/${folderForm}/${folderBuild}/index.html`}}
                onNavigationStateChange={this.onNavigationStateChange}
                onLoad={this.onLoad}
                allowUniversalAccessFromFileURLs
                allowFileAccess
                onMessage={this.onMessage}
            />
        );
    }
}

WebForm.propTypes = {
    spreadWebView: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
        flowStatus: state.FormRender.flowStatus,
        dataNextStep: state.FormRender.nextStep.data,
        pathNextStep: state.FormRender.nextStep.path,
        errorNextStep: state.FormRender.nextStep.error,
        formPath: state.FormRender.formPath,
        formDataPath: state.FormRender.formDataPath,
        stepData: state.FormRender.stepData,
        fileInput: state.FormRender.fileInput,
        fileVersions: state.FormRender.fileVersions,
        fileUploaded: state.FormRender.dataFile,
        location: state.FormRender.location
    }),
    mapDispatchToProps = dispatch => ({
        onMessage: (data) => {
            if (Actions.formRender.bridge[data.type]) {
                dispatch(Actions.formRender.bridge[data.type](data));
            }
        },
        finishLoadWebView: () => {
            dispatch(Actions.screens.all.disableItemList(true));
        },
        clearWebForm () {
            dispatch(Actions.formRender.nextStep.reset());
        }
    });

export default connect(mapStateToProps, mapDispatchToProps)(WebForm);
