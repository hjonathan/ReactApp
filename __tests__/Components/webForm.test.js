import 'react-native';
import React from 'react';
import WebFormComponent from '../../App/Components/WebForm';
import {WebForm} from '../../App/Components/WebForm';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import Data from "../../App/Model/Data";
import {CASE} from "../../App/Libs/Const";

describe('Test of Web form component', () => {
    const initialState = {
            FormData: {
                server: {
                    error: null
                }
            },
            FormRender: {
                routed: {
                    isRouted: false,
                    users: []
                },
                nextStep: {
                    error: null
                },
                getNextStep: "",
                flowStatus: "",
                nextStepPath: "",
                formPath: "",
                formDataPath: "",
                stepData: "",
                fileVersions: {
                    fieldID: "image1234",
                    appDocUid: "Doc000122",
                    response: "[]",
                    success: true
                }
            },
            Cases: {
                error: {},
                assignment: {
                    info: [
                        {
                            test: ""
                        }
                    ],
                    error: null
                },
                routedCase: {
                    info: null,
                    error: null
                },
                routeName: CASE.INBOX
            }
        },
        webView = {
            injectJavaScript: jest.fn()
        },
        payload = {app_uid: "WF133"},
        spreadWebView = jest.fn(),
        onMessage = jest.fn(),
        finishLoadWebView = jest.fn(),
        clearWebForm = jest.fn();

    const mockStore = configureStore(),
        nextProps = {
            errorNextStep: {}
        },
        nextPropsView = {
            flowStatus: 'continueStep'
        };
    let store, container, view;

    beforeEach(() => {
        store = mockStore(initialState);
        container = shallow(
            <WebFormComponent
                store={store}
                spreadWebView={spreadWebView}
                onMessage={onMessage}
                finishLoadWebView={finishLoadWebView}
                clearWebForm={clearWebForm}
            />
        );
        view = shallow(
            <WebForm
                payload={payload}
                store={store}
                spreadWebView={spreadWebView}
                onMessage={onMessage}
                finishLoadWebView={finishLoadWebView}
                clearWebForm={clearWebForm}
            />
        );
    });

    it('render Item correctly', () => {
        expect(toJson(container)).toMatchSnapshot();
    });

    it('Render the connected(SMART) component', () => {
        expect(container.length).toEqual(1)
    });

    it('mapStateToProps', () => {
        const data = {
            type: "continueStep"
        };
        expect(container.props('children').spreadWebView()).toMatchSnapshot();
        expect(container.props('children').onMessage(data)).toMatchSnapshot();
        expect(container.props('children').finishLoadWebView()).toMatchSnapshot();
        expect(container.props('children').clearWebForm()).toMatchSnapshot();
    });

    it("componentWillReceiveProps", () => {
        view.instance().webView = webView;
        expect(view.instance().componentWillReceiveProps(nextProps)).toMatchSnapshot();
    });

    it("componentDidUpdate", () => {
        view.instance().webView = webView;
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "continueStep"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "loadForm"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "hideFieldLoading"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "setFormData"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "setFiles"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "setFileVersions", fileVersions: initialState.FormRender.fileVersions});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "setLocation"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "getFormData"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
        view.setProps({flowStatus: "setCacheLibraryMap"});
        expect(view.instance().componentDidUpdate(nextPropsView, {})).toMatchSnapshot();
    });

    it('onMessage function', () => {
        const data = {
            nativeEvent: {
                data: '{"type":"","data":""}'
            }
        };
        expect(view.instance().onMessage(data)).toMatchSnapshot();
    });

    it('onLoad function', () => {
        view.setProps({
            state: {
                Settings: {
                    error: {
                        error: "",
                        error_description: ""
                    },
                    server: {
                        url: "https://poc-1.processmaer.net",
                        workspace: "workflowpoc1"
                    }
                },
                Oauth: {credentials: {token: "token"}},
                Net: {isConnected: true}
            }
        });
        view.instance().webView = webView;
        expect(view.instance().onLoad({nativeEvent: {}})).toMatchSnapshot();
    });

    it('onLoad function', () => {
        view.setProps({
            state: {
                Settings: {
                    error: {
                        error: "",
                        error_description: ""
                    },
                    server: {
                        url: "https://poc-1.processmaer.net",
                        workspace: "workflowpoc1"
                    }
                },
                Oauth: {credentials: {token: "token"}},
                Net: {isConnected: true}
            }
        });
        view.instance().webView = webView;
        Data.saveCaseData({params: {
                app_uid: "WF133",
                del_index: "3",
                dyn_uid: "4783964",
                status: CASE.STATUS.WORKING
            },
            response: {var001: 'value001'}
        }, initialState);
        expect(view.instance().onLoad({nativeEvent: {}})).toMatchSnapshot();
    });

    it('render ref function', () => {
        let webView = view.instance().render();
        expect(webView.ref()).toMatchSnapshot();
    });
});
