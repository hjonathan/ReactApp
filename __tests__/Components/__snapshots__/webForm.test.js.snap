// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of Web form component componentDidUpdate 1`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 2`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 3`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 4`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 5`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 6`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 7`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 8`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 9`] = `undefined`;

exports[`Test of Web form component componentDidUpdate 10`] = `undefined`;

exports[`Test of Web form component componentWillReceiveProps 1`] = `undefined`;

exports[`Test of Web form component mapStateToProps 1`] = `undefined`;

exports[`Test of Web form component mapStateToProps 2`] = `undefined`;

exports[`Test of Web form component mapStateToProps 3`] = `undefined`;

exports[`Test of Web form component mapStateToProps 4`] = `undefined`;

exports[`Test of Web form component onLoad function 1`] = `undefined`;

exports[`Test of Web form component onLoad function 2`] = `undefined`;

exports[`Test of Web form component onMessage function 1`] = `undefined`;

exports[`Test of Web form component render Item correctly 1`] = `
<WebForm
  clearWebForm={[Function]}
  errorNextStep={null}
  fileVersions={
    Object {
      "appDocUid": "Doc000122",
      "fieldID": "image1234",
      "response": "[]",
      "success": true,
    }
  }
  finishLoadWebView={[Function]}
  flowStatus=""
  formDataPath=""
  formPath=""
  onMessage={[Function]}
  spreadWebView={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  stepData=""
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`Test of Web form component render ref function 1`] = `undefined`;
